common/expected.hpp

```diff
@@ -0,0 +1,123 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <cassert>
+#include <string>
+#include <type_traits>
+
+#ifndef EXPECTED_HPP
+#define EXPECTED_HPP
+
+template <typename ErrorCodeType>
+class Error {
+public:
+	static_assert(
+		std::is_enum<ErrorCodeType>::value, "Error requires an enum type for error codes");
+	ErrorCodeType ec;
+	std::string msg;
+
+	Error(ErrorCodeType ec, std::string msg) {
+		this->ec = ec;
+		this->msg = msg;
+	};
+	Error(const Error &e) {
+		this->ec = e.ec;
+		this->msg = e.msg;
+	};
+};
+
+template <typename ExpectedType, typename ErrorType>
+class Expected {
```

Comment:
I would actually vote for having it uppercase to separate it tbh

Reply:
Here's one way I considered setting it up to ensure we stay compatible:

```c++
#if __cplusplus < 202303L
namespace mender_common {
class expected {
    ...
}
} // namespace mender_common
namespace std {
using mender_common::expected;
}
#endif
```

Which would enable our (limited) version to be exported as `std::expected` if, and only if, it is not available already. This means we could compile with a more recent compiler to check that we are still compatible with the standard.

It's similar to how I was planning to enable `make_unique`:

```c++
#if __cplusplus < 201402L
#include <boost/smart_ptr/make_unique.hpp>

namespace std {
using boost::make_unique;
}
#endif
```

Reply:
> Which would enable our (limited) version to be exported as std::expected if, and only if, it is not available already. This means we could compile with a more recent compiler to check that we are still compatible with the standard.

It also means that most of our builds would use the `std::expected` implementation and potential bugs in our own implementation may slip through. Is there a good reason to prefer the `std::expected` over our own implementation if it's available?

Reply:
No strong reason, it's mostly about code liability, so to speak. I assume their version is better, whether it is more efficient, catching more corner case bugs, or whatever. And long term we want to shed as much code as possible that we don't want to maintain ourselves, so it would be great if this file could just be removed in the future and it wouldn't affect the build in a significant way.

Compare it to CFEngine `printf` implementation. It's good that we have it for compatibility reasons, but we still would like to use the system version if we can.

You have a valid point about bugs in our own implementation, but since C++23 is not even out yet, I think we have plenty of time to verify that before we compile with any compiler where the real `std::expected` is actually enabled.

Reply:
It also means we won't be able to fully switch to `std::expected` any time soon, if ever.

Reply:
Not for a long time, no. 🙂

It's not that important, so if you don't think it's a good idea, feel free to just drop it.

Reply:
I'll stick to `Expected` now. We can run `git grep 'Expected' | xargs sed -i ...` any time :)